### DISCUSSION TITLE
error while interacting with dapp on mobile

### DISCUSSION THREAD

**Discussion by Rickshanelorie** [WalletConnect/walletconnect-monorepo#3305] [Developer Support] [answered]
i get the error on mobile only but on PC it's working fine inpage.js:1 MetaMask - RPC Error: The requested account and/or method has not been authorized by the user. {code: 4100, message: 'The requested account and/or method has not been authorized by the user.'} while using metamask or trust wallet

here's my code 

======================================================================================

"import {
  EthereumClient,
  w3mConnectors,
  w3mProvider,
  WagmiCore,
  WagmiCoreChains,
  WagmiCoreConnectors,
} from "https://unpkg.com/@web3modal/ethereum@2.6.2";
import { Web3Modal } from "https://unpkg.com/@web3modal/html@2.6.2";

// 0. Import wagmi dependencies
const { bscTestnet, bsc } = WagmiCoreChains;
const { configureChains, createConfig } = WagmiCore;

// 1. Define chains
const chains = [bscTestnet, bsc];
const projectId = "#";

// 2. Configure wagmi client
const { publicClient } = configureChains(chains, [w3mProvider({ projectId })]);
const wagmiConfig = createConfig({
  autoConnect: true,
  connectors: [
    ...w3mConnectors({ chains, version: 2, projectId }),
    new WagmiCoreConnectors.CoinbaseWalletConnector({
      chains,
      options: {
        appName: "#",
      },
    }),
  ],
  publicClient,
});

// 3. Create ethereum and modal clients
const ethereumClient = new EthereumClient(wagmiConfig, chains);
export const web3Modal = new Web3Modal(
  {
    projectId,
    themeVariables: {
      "--w3m-accent-color":
        "background: linear-gradient(157.81deg, #DEF9FA -43.27%, #BEF3F5 -21.24%, #9DEDF0 12.19%, #7DE7EB 29.82%, #5CE1E6 51.94%, #33BBCF 90.29%);",
    },
    walletImages: {
      safe:
        "https://pbs.twimg.com/profile_images/1566773491764023297/IvmCdGnM_400x400.jpg",
    },
  },
  ethereumClient
);

const getAccountAndHandle = async () => {
  try {
    account = await ethereumClient.getAccount();
    console.log(account);
  } catch (err) {
    console.error(err);
  }
};

ethereumClient.watchAccount(async (accounts) => {
  console.log("Accounts:", accounts);
  await getAccountAndHandle();
});

getAccountAndHandle();"

**Comment by Rickshanelorie** [comment]
@glitch-txs

**Reply by Rickshanelorie** [reply]
@glitch-txs brother can you plz help me here ?>??

**Comment by Rickshanelorie** [comment]
why everyone gets an reply but mee not???

**Comment by C00K1E-dev** [comment]
i have the same isue on pc everything works fine but on mobile it does not work ... the state of the connection even tho i am connected when i try the button to mint an NFT  the connected state is not recognized ... its the same for view my nfts button .... but this is only mobile version the pc version works fine

**Reply by Rickshanelorie** [reply]
YES you have to use the wallet browser on phone to work overall you facing the the same issue as me and no one from the wallet connect team want to reply i don't know why

**Reply by Rickshanelorie** [reply]
send my you're telegram brother

**Reply by C00K1E-dev** [reply]
@D3y0o for telegram black and white selfie profile picture

**Comment by glitch-txs** [comment]
Hi @Rickshanelorie ! It seems like you're trying to request the wallet's account before establishing a connection, do you have a public repo or could provide a complete example on this issue?

Also would recommend using backticks when pasting code on GitHub, would helps us a lot :)
https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Reply by Rickshanelorie** [reply]
hey @glitch-txs thanks for the reply , here you go https://github.com/Rickshanelorie/walletconnectissue/tree/main

**Reply by Rickshanelorie** [reply]
@glitch-txs any updates brother

**Reply by glitch-txs** [reply]
This happens when you scan the QR code with MetaMask? worked fine for me, also could you try on our [examples](https://github.com/WalletConnect/web3modal-examples)? They are very minimal which makes it easier to debug

**Reply by Rickshanelorie** [reply]
so you were able to interact with the dapp after scanning the qr code without getting the error MetaMask - RPC Error: The requested account and/or method has not been authorized by the user. {code: 4100, message: 'The requested account and/or method has not been authorized by the user.'}

**Reply by Rickshanelorie** [reply]
I mean the dapp work on the wallet browser

**Reply by Rickshanelorie** [reply]
but I want it after I scan the qr code to work for me and being able to interact

**Reply by glitch-txs** [reply]
sorry didn't see the messages, yes I was able to connect

**Reply by Rickshanelorie** [reply]
my only problem is when i connect with qr code i get the error script.js:203 s: Returned error: The requested account and/or method has not been authorized by the user other then that no issue with my dapp if i use it with browser extension no issue mobile wallet browser no issue too this error when i try to write any function when am connected using the qr code i hope you get it

**Reply by glitch-txs** [reply]
Try clearing localStorage

**Reply by Rickshanelorie** [reply]
done still getting 

"inpage.js:1 MetaMask - RPC Error: The requested account and/or method has not been authorized by the user. {code: 4100, message: 'The requested account and/or method has not been authorized by the user.'}" i don't get it if the account has not been authorized  why it work when i use the wallet browser or the browser extension

**Reply by Rickshanelorie** [reply]
@glitch-txs

**Reply by glitch-txs** [reply]
I will try again later, what was your MetaMask version and mobile? Also did you test connecting on our example Dapps?

**Reply by Rickshanelorie** [reply]
yes i did test on dapp exmples, am using android and latest verison of metamask

**Reply by Rickshanelorie** [reply]
also here's my dapp "https://atlasfund.app/dapp", also if you can send any bscscantestnet address ill send you some of our test tokens to test

**Reply by glitch-txs** [reply]
I'm able to connect with MetaMask on android
![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/66949816/f1ab15cf-fa2e-43e8-8fdd-ad999409cc36)

**Reply by glitch-txs** [reply]
If you're working on testnet keep in mind some wallet don't support testnet for security reason, also there's a [current issue](https://github.com/MetaMask/metamask-mobile/issues/6655) with chain switching on MetaMask that's being worked on

**Reply by Rickshanelorie** [reply]
yes but if you try to approve now for example you will get the error if you're connected using qr code but if you connected using the browser extension or the wallet browser the error will not happen

**Reply by glitch-txs** [reply]
can you show me how are you calling approve function?

**Reply by Rickshanelorie** [reply]
brother it's not the approve function issue like I told you if connect using the wallet browser or the browser extension you will be able to interact with the dapp no issue but if you connect using the qr code you will get the error in the console for any function you can try visiting the site and try connecting using the extension first and try to approve or any function in the site you will get no issue then try using qr code

**Reply by C00K1E-dev** [reply]
@glitch-txs I have the same problem mate I'm just trying a mint button it works perfectly on pc just like @Rickshanelorie it works perfectly on mobile in metamask wallet browser just like  @Rickshanelorie but if open in safari browser on iOS or Google browser on iOS or android phone also Google browser it doesn't work anymore I cannot mint ... what hooks should I use Wagmi ? Web3 ? Should I use provider ? Tried a bunch of stuffs so the web3modal calls the metamask app to connect and I am connected but the mint button doesn't call metamask to sign the transaction any ideas ?

**Reply by glitch-txs** [reply]
here's a minimal reproducible example for mainnet, so far it worked for me on MetaMask mobile (the transaction shows in the wallet)
https://github.com/glitch-txs/example-tx

Please let's simplify the examples like this so we can help you better 👍

**Reply by C00K1E-dev** [reply]
> here's a minimal reproducible example for mainnet, so far it worked for me on MetaMask mobile (the transaction shows in the wallet)
> 
> https://github.com/glitch-txs/example-tx
> 
> 
> 
> Please let's simplify the examples like this so we can help you better 👍 

Thank you for this example I will try a bit later to reproduce

**Reply by Rickshanelorie** [reply]
let me know with the updates bro

**Reply by C00K1E-dev** [reply]
@glitch-txs  tried the example you provided...  you are importing WagmiCoreChains from @web3moda/ethereum  that package doesn't contain WagmiCoreChains     ... you mean configureChains from  @wagmi/core?

**Reply by C00K1E-dev** [reply]
https://github.com/WalletConnect/walletconnect-monorepo/assets/122067978/82eb6843-f5d2-4063-a33c-268e612960ac
perhaps you can see the problem better in video ...

**Reply by Rickshanelorie** [reply]
just tried you're exmple still facing the same issue with the qr code

**Reply by glitch-txs** [reply]
@C00K1E-dev seems like you're checking for `window.ethereum` which doesn't exists on mobile browsers (unless you are in a wallet's browser like shown in the video). WalletConnect doesn't require window.ethereum in order to work with mobile wallets.

**Reply by C00K1E-dev** [reply]
@glitch-txs  Just replied to you in my open issue 
I have invited you to the repository you can see the structure. Perhaps you can give some advice and tips tried quite a few options this last days non of em worked ! Thank you in advance

**Reply by Rickshanelorie** [reply]
so what's the fix because it seems like we tried everything on the internet I even shared my whole code with you

**Reply by C00K1E-dev** [reply]
> @C00K1E-dev seems like you're checking for `window.ethereum` which doesn't exists on mobile browsers (unless you are in a wallet's browser like shown in the video). WalletConnect doesn't require window.ethereum in order to work with mobile wallets.

@glitch-txs  Could you please tell what should I use then if not 'window.ethereum' ? Thank you!

**Reply by Rickshanelorie** [reply]
exactly

**Reply by C00K1E-dev** [reply]
> exactly 

It's like they keep it a secret 😂😂

**Reply by glitch-txs** [reply]
could you show me the error in my example? I was actually able to tx there

**Reply by C00K1E-dev** [reply]
> could you show me the error in my example? I was actually able to tx there

Nevermind I have managed to make it work I can mint now from mobile  browsers also ... so the thing was ethereumClient was rendered higher  in the structure of my app and the mint button could not get the state of the connected account I passed ethereumClient to my component I then used Wagmi to get the connected account and it works now

**Reply by Rickshanelorie** [reply]
hey @glitch-txs for me it still not fixed can you plz take a look at the repo and tell me what's wrong btw i tested on mainnet still same error 

https://github.com/Rickshanelorie/exmp/blob/main/exmple.js

thanks for you're time

**Reply by glitch-txs** [reply]
Hi Rick unfortunately I'm pretty busy to go through the whole code, it would be easier if we could simplify this in a minimal reproducible example: https://stackoverflow.com/help/minimal-reproducible-example

**Reply by Rickshanelorie** [reply]
hey @glitch-txs that's the most simplify example i could put it just the rpc configuration and the wallet connect code and the approve function as an exmple, i would appreciate it if you can take a small look

**Reply by glitch-txs** [reply]
I will take a look, thanks 🙏

**Reply by Rickshanelorie** [reply]
@glitch-txs any updates sir ?

**Reply by glitch-txs** [reply, marked as answer]
Hi the error seems to come from web3.js, you're trying to trigger a transaction without the signer. Web3modal works with wagmi instead of web3.js, you should use this instead:

https://wagmi.sh/core/actions/writeContract

**Reply by Rickshanelorie** [reply]
hey thanks for the response so after i write it using wagmi the qr code should work right ?

**Reply by Rickshanelorie** [reply]
@glitch-txs oh thanks it just worked
